core/templates/pages/classroom-admin-page/modals/create-new-classroom-modal.component.ts

```diff
@@ -0,0 +1,174 @@
+
+// Copyright 2022 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Create new classroom modal.
+ */
+
+import { Component } from '@angular/core';
+import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
+import { AppConstants } from 'app.constants';
+import { ConfirmOrCancelModal } from 'components/common-layout-directives/common-elements/confirm-or-cancel-modal.component';
+import { ClassroomBackendApiService } from '../../../domain/classroom/classroom-backend-api.service';
+
+
+@Component({
+  selector: 'oppia-create-new-classroom-modal',
+  templateUrl: './create-new-classroom-modal.component.html'
+})
+export class CreateNewClassroomModalComponent
+  extends ConfirmOrCancelModal {
+  constructor(
+    private classroomBackendApiService: ClassroomBackendApiService,
+    private ngbActiveModal: NgbActiveModal
+  ) {
+    super(ngbActiveModal);
+  }
+
+  existingClassroomNames: string[] = [];
+  newClassroomName: string = '';
+  newClassroomId: string = '';
+  newClassroomUrlFragment: string = '';
+  creatingNewClassroom: boolean = false;
+
+  classroomNameExceedsMaxLen: boolean = false;
+  emptyClassroomName: boolean = false;
+  duplicateClassroomName: boolean = false;
+  classroomNameIsValid: boolean = false;
+
+  classroomUrlFragmentExceedsmaxLen: boolean = false;
+  emptyClassroomUrlFrgament: boolean = false;
+  duplicateClassroomUrlFragment: boolean = false;
+  urlFragmentRegexMatched: boolean = true;
+  classroomUrlFragmentIsValid: boolean = false;
```

Comment:
True, there is some code repetition my view regarding this is – 
In the classroom admin page component there are several properties apart from classroom name and URL fragment. So wouldn't it cause some confusion there because (after introducing a service) some properties are initialized and validated by an external service while some of the properties are initialized locally in the component?

Looking forward to your suggestion.
Thank you

Reply:
I think we could still separate the common logic to a service, it may be that validation would be different in the two scenarios e.g. in the creation case, the validation may be only for the required URL and name fields whereas in the other component it will validate more fields.


I would recommend creating separate interfaces for a "new classroom" and an "existing classroom". A new classroom object can be constructed with a minimal set of properties whereas the existing classroom will be more exhaustive. Both the interfaces will contain a validate method, which will allow us to create separate classes with different validate  methods. We can define a type "ClassroomData" that will be the union of "NewClassroom" and "ExistingClassroom" -- this is the typescript way of implementing inheritance. The service will then deal only with ClassroomData. Calling the validate method of ClassroomData in the service will validate the relevant classroom object. Here's a sample implementation of what I taked about:


```
interface NewClassroom {
    name: string;
    url: string;
    validate: () => boolean
}

interface ExistingClassroom extends NewClassroom {
    // The extra properties that an existing classroom may contain.
    extra1: number;
    extra2: number;
    validate: () => boolean
}

// Union type
type ClassroomData = NewClassroom | ExistingClassroom;

class NewClassroomData implements NewClassroom {
    name: string;
    url: string;
    constructor(name: string, url: string) {
        this.name = name;
        this.url = url;
    }
    validate(): boolean {
        return !!(this.name && this.url)
    }
}

class ExistingClassroomData implements ExistingClassroom {
    name: string;
    url: string;
    extra1: number;
    extra2: number;
    constructor(name: string, url: string, extra1: number, extra2: number) {
        this.name = name;
        this.url = url;
        this.extra1 = extra1;
        this.extra2 = extra2;
    }
    validate(): boolean {
        return !!(this.name && this.url && this.extra1 && this.extra2)
    }
}

const newClass = new NewClassroomData('new', '3');
const existingClass = new ExistingClassroomData('exist', 'url', 0, 1);


// In the service, we only use the type "ClassroomData". 
const validateClassroom = (classroom: ClassroomData) => {
    // Calling validate will do a dynamic dispatch and call the
    // validation method in the relevant classroom object.
    return classroom.validate();
}

// Outputs true
console.log(validateClassroom(newClass));

// Outputs false because extra1 is 0
console.log(validateClassroom(existingClass));
```

Reply:
Thank you @kevintab95 this code explanation really helped a lot.

I have created a new file "classroom-admin.model" which contains the classroom-related fields, variables, and methods.
The model also helped me to reduce duplicate code. 
Please take a look at the updated code, once you are free.

Thank you